pkg/descheduler/strategies/topologyspreadyconstraint.go

```diff
@@ -0,0 +1,150 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package strategies
+
+import (
+	"fmt"
+	"math"
+
+	"github.com/golang/glog"
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	clientset "k8s.io/client-go/kubernetes"
+
+	"github.com/kubernetes-incubator/descheduler/cmd/descheduler/app/options"
+	"github.com/kubernetes-incubator/descheduler/pkg/api"
+	"github.com/kubernetes-incubator/descheduler/pkg/descheduler/evictions"
+	podutil "github.com/kubernetes-incubator/descheduler/pkg/descheduler/pod"
+)
+
+func TopologySpreadConstraint(ds *options.DeschedulerServer, strategy api.DeschedulerStrategy, policyGroupVersion string, nodes []*v1.Node, nodePodCount nodePodEvictedCount) {
+	if !strategy.Enabled {
+		return
+	}
+
+	fmt.Printf("Found following parameters for TopologySpreadConstraint %v\n", strategy)
+	for _, topoConstraints := range strategy.Params.NamespacedTopologySpreadConstraints {
+		glog.Infof("Found Topo Constraints %v\n", topoConstraints)
+		evictPodsViolatingSpreadConstraints(ds.Client, policyGroupVersion, nodes, ds.DryRun, nodePodCount, topoConstraints)
+	}
+
+}
+
+// AntiAffinityTerm's topology key value used in predicate metadata
+type topologyPair struct {
+	key   string
+	value string
+}
+
+type podSet map[*v1.Pod]struct{}
+
+type topologyPairSet map[topologyPair]struct{}
+
+// finnd all nodes
+// find all pods on the nodes
+// find the pods that match the namespace
+// find the constraints topokey , use this topokey to finds the node topokey value
+// add the pod with key as the combination of topokey and its value
+// find the min value of pods with any key
+// iterate through all keys and diff currentPods -minPods <=maxSkew
+// if diff > maxSkew, select a pod in the current bucket for eviction
+//
+func evictPodsViolatingSpreadConstraints(client clientset.Interface, policyGroupVersion string, nodes []*v1.Node, dryRun bool, nodePodCount nodePodEvictedCount,
+	constraint api.NamespacedTopologySpreadConstraint) {
+
+	if len(constraint.TopologySpreadConstraints) != 1 {
+		glog.V(1).Infof("We currently only support 1 topology spread constraint per namespace")
+		return
+	}
+	topologyPairToPods := make(map[topologyPair]podSet)
+	for _, node := range nodes {
+		glog.V(1).Infof("Processing node: %#v\n", node.Name)
+		pods, err := podutil.ListEvictablePodsOnNode(client, node, false)
```

Comment:
I believe clientset is able to filter pods by namespace. We can pass in `namespace` hence don't need to check if `pod.Namespace != constraint.Namespace` later.

Reply:
Listing pods for each node will cause N API requests (N=len(nodes)). It's better to change the logic to:

```
for each topologySpreadConstraint
    pre-fetch all qualified pods globally (or use similar cache in descheduler if appropriate)
    process the pods list:
	(1) filter out the pods whose node doesn't have needed topologyKey
	(2) so that we can get a map[nodeName]podSet, and also we know the minimum match number
	for nodeName, podSet := range map[nodeName]podSet
		needEvictNum := len(podSet) - minMatch - maxSkew
		if needEvictNum > 0
			evict needEvictNum pods from this Node
			Note: this math is sort of brute force, we can come up with better math later.
			For example, for a 5/1/3 cluster, and maxSkew is 1; the brute force math above will
			evict 3/0/1 pods from each topology, but if we consider the math "dynamically", we
			should only evict 2/0/0 pods, so that eventually it can become 3/3/3.
```